Userland/Libraries/LibGUI/FilePicker.cpp

```diff
@@ -113,11 +117,41 @@ FilePicker::FilePicker(Window* parent_window, Mode mode, StringView filename, St
         set_path(m_location_textbox->text());
     };
 
-    auto open_parent_directory_action = Action::create(
-        "Open parent directory", { Mod_Alt, Key_Up }, Gfx::Bitmap::try_load_from_file("/res/icons/16x16/open-parent-directory.png"sv).release_value_but_fixme_should_propagate_errors(), [this](Action const&) {
-            set_path(DeprecatedString::formatted("{}/..", m_model->root_path()));
-        },
-        this);
+    auto* file_types_filters_combo = widget.find_descendant_of_type_named<GUI::ComboBox>("allowed_file_type_filters_combo");
+
+    if (m_allowed_file_types.has_value()) {
+        for (auto& filter : *m_allowed_file_types) {
+            if (!filter.extensions.has_value()) {
+                m_allowed_file_types_names.append(filter.name);
+                continue;
+            }
+
+            StringBuilder extension_list;
+            extension_list.join("; "sv, *filter.extensions);
+            m_allowed_file_types_names.append(DeprecatedString::formatted("{} ({})", filter.name, extension_list.to_deprecated_string()));
+        }
+
+        file_types_filters_combo->set_model(*GUI::ItemListModel<DeprecatedString, Vector<DeprecatedString>>::create(m_allowed_file_types_names));
+        file_types_filters_combo->on_change = [this](DeprecatedString const&, GUI::ModelIndex const& index) {
+            m_model->set_allowed_file_extensions((*m_allowed_file_types)[index.row()].extensions);
+        };
+        file_types_filters_combo->set_selected_index(0);
+    } else {
+        auto* file_types_filter_label = widget.find_descendant_of_type_named<GUI::Label>("allowed_file_types_label");
+        auto& spacer = file_types_filter_label->parent_widget()->add<GUI::Widget>();
+        spacer.set_fixed_height(22);
+        file_types_filter_label->remove_from_parent();
+
+        file_types_filters_combo->parent_widget()->insert_child_before(GUI::Widget::construct(), *file_types_filters_combo);
```

Comment:
I'm not sure about this pattern, but if it's been used elsewhere then no comment on the pattern itself.

One question though: Shouldn't this new child have fixed height of 22 as well to preserve the old layout before the combo box existed? Or does the filter types label take care of that?

Reply:
That code reverts the GUI to how it looked before this PR, so it removes the "Files of Type:" label with an empty widget(reverts [this](https://github.com/SerenityOS/serenity/pull/16363/files#diff-e383b1da93f7b5cd5743ced5fafe25792fb68a7f97ee2a808ba34a9ff08e7448L32)) which requires a fixed height of 22 to look correct and it also replaces the filter combo box with an empty widget(reverts [this](https://github.com/SerenityOS/serenity/pull/16363/files#diff-e383b1da93f7b5cd5743ced5fafe25792fb68a7f97ee2a808ba34a9ff08e7448L85)), which in the original GML was a `GUI::Layout::Spacer` but I couldn't find a way to insert a spacer before or after a widget so I went with a widget and everything looks fine even when height isn't specified (see image below).

![2022-12-15_19-11](https://user-images.githubusercontent.com/18418792/207937631-6dac87f5-676b-4fe5-9195-349eddb72b6c.png)

About this pattern, I'm not sure, it's the only way I can think of, if I just remove the label and combo box the GUI will be wrong (see image below),

![2](https://user-images.githubusercontent.com/18418792/207939950-d555dc3a-44ef-4a60-ae46-a89105f5cd7e.png)

Reply:
Makes sense, thanks for the explanation!